### PR TITLE
Implement MOI wrapper

### PR DESCRIPTION
**Pull request by blegat**
- [x] Infeasible NLP
- [x] Heuristic solution with incumbent found by NLP
- [x] Quadratic constraints
- [x] Quadratic objective

Requires https://github.com/jump-dev/Ipopt.jl/issues/211
Closes https://github.com/JuliaOpt/Pavito.jl/issues/7

### CODE REVIEW

**Comment by mlubin** [comment]
Wow, how long did this take?

**Comment by blegat** [comment]
I started after this commit: https://github.com/lanl-ansi/Alpine.jl/pull/149/commits/bb3c7a8970d91ba685420aa58cf29cfa948089f5 which was 1:24 pm. This this commit: https://github.com/JuliaOpt/Pavito.jl/pull/21/commits/ca6a13e6cbd9734702babba460779ae868275ee0 is at 8:10 pm, it took 6h46.
With MOI, the code is simpler as you don't need to convert the problem into any specific format, you just forward everything to the inner solvers.

**Comment by ccoffrin** [comment]
Re tests, I would also recommend giving https://github.com/jump-dev/MINLPTests.jl a try.

**Comment by blegat** [comment]
I have moved remaining improvements #22, #23, #24 and #25. I prefer starting with a minimal port to MOI. Let me know whether you have any comments.

**Comment by ccoffrin** [comment]
I would recommend that this solver pass the MINLP tests before a version is tagged, they are designed to test the very basic NLP API functionality, similar to the MOI tests; if you prefer to do it in a seperate PR that's fine by me.

**Comment by blegat** [comment]
It would be a separate PR but we can wait for it before tagging. This waiting period can also be used for people to try out the master version and report regressions.

File: src/MOI_wrapper.jl [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    rel_gap::Float64                        # Relative optimality gap termination condition
+    mip_solver_drives::Union{Nothing, Bool} # Let MILP solver manage convergence ("branch and cut")
+    mip_solver                              # MILP solver
+    cont_solver                             # Continuous NLP solver
+
+    mip_optimizer::Union{Nothing, MOI.ModelLike}
+    cont_optimizer::Union{Nothing, MOI.ModelLike}
+    inf_optimizer::Union{Nothing, MOI.ModelLike}
+
+    θ::Union{Nothing, MOI.VariableIndex}
```
Comment (mlubin) [inline src/MOI_wrapper.jl:26]: What's θ?

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+# Pavito solver
+mutable struct Optimizer <: MOI.AbstractOptimizer
+    log_level::Int                          # Verbosity flag: 0 for quiet, higher for basic solve info
+    timeout::Float64                        # Time limit for algorithm (in seconds)
+    rel_gap::Float64                        # Relative optimality gap termination condition
+    mip_solver_drives::Union{Nothing, Bool} # Let MILP solver manage convergence ("branch and cut")
+    mip_solver                              # MILP solver
+    cont_solver                             # Continuous NLP solver
+
```
Comment (mlubin) [inline src/MOI_wrapper.jl:21]: Most of these fields could use descriptive comments.

File: src/MOI_wrapper.jl [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    inf_optimizer::Union{Nothing, MOI.ModelLike}
+
+    θ::Union{Nothing, MOI.VariableIndex}
+    mip_variables::Vector{MOI.VariableIndex}
+    cont_variables::Vector{MOI.VariableIndex}
+    inf_variables::Vector{MOI.VariableIndex}
+    nl_slack_variables::Union{Nothing, Vector{MOI.VariableIndex}}
+    quad_less_than_slack_variables::Union{Nothing, Vector{MOI.VariableIndex}}
+    quad_greater_than_slack_variables::Union{Nothing, Vector{MOI.VariableIndex}}
+    quad_less_than_inf_con::Union{Nothing, Vector{MOI.ConstraintIndex{SQF, MOI.LessThan{Float64}}}}
```
Comment (mlubin) [inline src/MOI_wrapper.jl:33]: This name reads like "quad less than infinity constraint"

**Comment by blegat** [comment]
One think I wanted to write down somewhere: there is one change of behavior of Pavito that might occur from before to after this PR. There was a but here:
https://github.com/JuliaOpt/Pavito.jl/blob/2d7d7e6948cfca795d54c1156a8fc14e3923e20b/src/algorithm.jl#L389
that was fixed, the `numnlconstr` should have been `numvars`.

**Pull request merged**